module/VuFind/src/VuFind/Db/Service/ShortlinksService.php

```diff
@@ -0,0 +1,178 @@
+     *
+     * @param string $path          Path to store in database
+     * @param string $hash          Hash of $path (generated in getGenericHash)
+     * @param int    $length        Minimum number of characters from hash to use for
+     *                              lookups (may be increased to enforce uniqueness)
+     * @param int    $maxHashLength The maximum allowed hash length
+     *
+     * @return string
+     */
+    public function saveAndShortenHash($path, $hash, $length, $maxHashLength)
+    {
+        // Validate hash length:
+        if ($length > $maxHashLength) {
+            throw new \Exception(
+                'Could not generate unique hash under ' . $maxHashLength
+                . ' characters in length.'
+            );
+        }
+        $shorthash = str_pad(substr($hash, 0, $length), $length, '_');
+
+        $queryBuilder = $this->entityManager->createQueryBuilder();
+        $queryBuilder->select('s')
+            ->from(Shortlinks::class, 's')
+            ->where('s.hash = :hash')
+            ->setParameter('hash', $shorthash);
+        $query = $queryBuilder->getQuery();
+        $results = $query->getResult();
+
+        // Brand new hash? Create row and return:
+        if (count($results) == 0) {
+            $shortlink = $this->createEntity();
+            $now = new \DateTime();
+            // Generate short hash within a transaction to avoid odd timing-related
+            // problems:
+            $this->entityManager->getConnection()->beginTransaction();
+            $shortlink->setHash($shorthash)
+                ->setPath($path)
+                ->setCreated($now);
+            try {
+                $this->entityManager->persist($shortlink);
+                $this->entityManager->flush();
+                $this->entityManager->getConnection()->commit();
+            } catch (\Exception $e) {
+                $this->logError('Could not save shortlink: ' . $e->getMessage());
+                $this->entityManager->getConnection()->rollBack();
+                throw $e;
+            }
+            return $shorthash;
+        }
+
+        // If we got this far, the hash already exists; let's check if it matches
+        // the path...
+        if (current($results)->getPath() === $path) {
+            return $shorthash;
+        }
+
+        // If we got here, we have encountered an unexpected hash collision. Let's
+        // disambiguate by making it one character longer:
+        return $this->saveAndShortenHash($path, $hash, $length + 1, $maxHashLength);
+    }
+
+    /**
+     * Resolve URL from Database via id.
+     *
+     * @param string $input   hash
+     * @param string $baseUrl Base URL of current VuFind site
+     *
+     * @return string
+     *
+     * @throws Exception
+     */
+    public function resolve($input, $baseUrl)
+    {
+        $queryBuilder = $this->entityManager->createQueryBuilder();
+        $queryBuilder->select('s')
+            ->from(Shortlinks::class, 's')
```

Comment:
`Shortlinks::class` should be changed to `$this->getEntityClass(Shortlinks::class)` to support custom overrides.